### ISSUE TITLE
Encoder/decoder return attention maps?

### ISSUE THREAD

**Issue by hadaev8** [lucidrains/x-transformers#5]
Would be nice for logging.

**Comment by lucidrains** [comment]
@hadaev8 sure!

**Comment by lucidrains** [comment]
@hadaev8 right after the softmax?

**Comment by hadaev8** [comment]
Yes

**Comment by lucidrains** [comment]
@hadaev8 got it, i'll get it done

**Comment by lucidrains** [comment]
@hadaev8 try version 0.6.3 with

```python
import torch
from x_transformers import TransformerWrapper, Encoder

model = TransformerWrapper(
    num_tokens = 20000,
    max_seq_len = 1024,
    attn_layers = Encoder(
        dim = 512,
        depth = 6,
        heads = 8
    )
)

x= torch.randint(1, 20000, (1, 1024))
logits, attn_maps = model(x)
print(attn_maps[0].shape) # (1, 8, 1024, 1024) attention map for first layer
```

**Comment by hadaev8** [comment]
@lucidrains 
Thanks, works like a charm.

But chance, do you know any tool for plotting attention maps nicely?
Like automatically making a grid for all heads and annotations.

**Comment by tomweingarten** [comment]
@hadaev8 Take a look at https://github.com/jessevig/bertviz, it should do what you want and if not it shouldn't be hard to modify it to

**Comment by hadaev8** [comment]
Thank you.
